WordPress/gutenberg
# [Dataviews] Table layout: Ensure focus is not lost on interaction

Resolves #57341.

## What?

This PR does ~three~ two relatively small things:
1. It changes the table header rendering behaviour, to always be present even when no data is available.
2. It makes sure that focus is suitably moved when a column is hidden from its header menu.
3. ~It persists stale data, to maintain visual consistency, which is rendered and marked as `inert` while fresh data is being fetched (sorting, pagination, etc).~


## Why?

When tables are sorted, filtered, or paginated, new data can be fetched. If this takes time to come back, the header is currently replaced with a "no results" message, until the fetch is complete. With the table header remaining in place, focus within header menus is persisted.

Separately, when columns are hidden from the column header menu, the menu is necessarily removed from the DOM, which leaves focus at the whim of the user agent. We need to be intentional about focus movement in this case.


## How?

1. The table header rendering is moved outside of the `hasData` check, so that it is always rendered, regardless of whether data exists.
2. A map is maintained of column headers, and suitable fallbacks, so that focus can be appropriately moved if a column is hidden.
3. ~A reference is maintained of previously rendered data, which is used when no current data is available.~


## Testing Instructions

1. Open the "Manage all pages" view in the new data views manager
2. Sort the data using a column header menu
   * Confirm that sorting different columns still works
   * Confirm that focus remains in the header menu
3. Paginate the data using the pagination controls
   * Confirm that data is still correctly updated
4. Filter the data (using a column header menu or filter controls)
   * Confirm that data is still correctly updated
   * Confirm that a 'No results' message is shown if no data matches the filter
5. Filter the data from the column header menu, to leave no results
   * Confirm that focus remains in the header menu
6. Remove a column using a column header menu
   * Confirm that focus is moved to the header menu in the previous column
   * Note: if removing the first column, focus should move to the next column instead


### Testing Instructions for Keyboard

The same interactions should be validated using the keyboard – focus should not be lost at any point.


## Screenshots or screencast

The table layout with no data – the header row is persisted.

![The DataViews Pages view, using the table layout. The data is filtered, with no results, and the message 'No results' is shown under the header row.](https://github.com/WordPress/gutenberg/assets/159848/721c2d19-c368-43ec-b0a0-4104d1006d5e)



Column header removal, showing focus being explicitly moved to a different column.

https://github.com/WordPress/gutenberg/assets/159848/b2aeb3e0-e345-4e7d-98d2-9d850585c314
